src/nft_contract/src/main.nr

```diff
@@ -50,118 +47,120 @@ pub contract NFT {
      * ======================================================== */
 
-    // Transfers token with `token_id` from public balance of message sender to a private balance of `to`.
     #[private]
-    fn transfer_to_private(to: AztecAddress, token_id: Field) {
-        let from = context.msg_sender();
-
-        let nft = NFT::at(context.this_address());
-
-        // We prepare the private balance increase.
-        let commitment = _prepare_private_balance_increase(to, &mut context, storage);
-
-        // At last we finalize the transfer. Usage of the `unsafe` method here is safe because we set the `from`
-        // function argument to a message sender, guaranteeing that he can transfer only his own NFTs.
-        nft._finalize_transfer_to_private_unsafe(from, token_id, commitment).enqueue(&mut context);
+    fn transfer_private_to_public(from: AztecAddress, to: AztecAddress, token_id: Field, nonce: Field) {
+        _validate_from_private(&mut context, from, nonce);
+        _remove_private_token_owner(storage.private_nfts.at(from), token_id);
+        NFT::at(context.this_address()).update_public_token_owner_internal(from, to, token_id, true).enqueue(
+            &mut context,
+        );
     }
 
-    /// Prepares an increase of private balance of `to` (commitment). The increase needs to be finalized by calling
-    /// `finalize_transfer_to_private` with the returned commitment.
     #[private]
-    fn prepare_private_balance_increase(to: AztecAddress) -> PartialNFTNote {
-        _prepare_private_balance_increase(to, &mut context, storage)
+    fn transfer_private_to_public_with_commitment(from: AztecAddress, to: AztecAddress, token_id: Field, commitment: PartialNFTNote, nonce: Field) -> PartialNFTNote {
+        _validate_from_private(&mut context, from, nonce);
+        _remove_private_token_owner(storage.private_nfts.at(from), token_id);
+        NFT::at(context.this_address()).update_public_token_owner_internal(from, to, token_id, true).enqueue(
+            &mut context,
+        );
+        let commitment = _initialize_transfer_commitment(&mut context, storage.private_nfts, from, to);
+        commitment
     }
 
     #[private]
-    fn transfer_in_private(from: AztecAddress, to: AztecAddress, token_id: Field, nonce: Field) {
+    fn transfer_private_to_private(from: AztecAddress, to: AztecAddress, token_id: Field, nonce: Field) {
         _validate_from_private(&mut context, from, nonce);
-
-        let nfts = storage.private_nfts;
-
-        let notes = nfts.at(from).pop_notes(NoteGetterOptions::new()
-            .select(NFTNote::properties().token_id, Comparator.EQ, token_id)
-            .set_limit(1));
-        assert(notes.len() == 1, "NFT not found when transferring");
-
-        let new_note = NFTNote::new(token_id, to);
-
-        nfts.at(to).insert(new_note).emit(encode_and_encrypt_note(&mut context, to, from));
+        let private_nfts = storage.private_nfts;
```

Comment:
curiously this doesn't "save up gas", so i wouldn't care on doing `storage.private_nfts.at(from)` in the following lines